Logging: New, improved view for a single log file

coreymckrill opened the pull request:
### Changes proposed in this Pull Request:

This provides a new view of a log file when you click on one in the https://github.com/woocommerce/woocommerce/pull/40662. It has several enhancements over the original log file view:

* Clickable line numbers. These provide permalinks to specific lines in the file.
* Formatting of some log entry elements. The severity level gets colorized so it's easier to see which log entries are more important, and the timestamp gets bolded so you can see where each log entry starts (since some entries are across multiple lines).
* Navigation links for switching between different "rotations" of a particular log.

This also refactors the File class a bit to use `WP_Filesystem` wherever possible instead of direct PHP filesystem calls.

Fixes #41143 

![Screen Shot 2023-11-02 at 16 08 37](https://github.com/woocommerce/woocommerce/assets/916023/c9629df7-826d-4b96-98fd-351d2d6a24b2)

### How to test the changes in this Pull Request:

#### Setup

1. In your test site's **wp-config.php** file, add this line: `define( 'WC_LOG_HANDLER', 'Automattic\\WooCommerce\\Internal\\Admin\\Logging\\LogHandlerFileV2' );`. This is what allows you to see the new log file views.
2. Install [this plugin](https://gist.github.com/coreymckrill/2968911c8b63c556e7787f2cdc65f966) on your test site for generating test log files. Probably easiest to copy the file into the mu-plugins directory.
3. Open two browser tabs: the Logs screen (WP Admin > WooCommerce > Status > Logs) and the WooCommerce Tools screen (WP Admin > WooCommerce > Status > Tools). On the Tools screen you should see the **Debug Log Generator** tool at the top of the list.

#### Tests

1. On the Logs screen, you should see a list table, which may already have some log files listed in it. Even if there are some, go ahead and generate some additional logs on the Tools screen so that we have enough different log files and sources. Generate two or three batches of test logs at once. You should then see that at least one of the logs, probably `debug-log-generator`, has more than one entry in the list table. These are "rotations" that are generated when a log file reaches a certain size.
2. Click on a log file that has one or more rotations. This should take you to the single file view, where you can see the contents of the file. There should be line numbers next to each line of the file.
3. Try clicking the line numbers. They are anchors, and the line you clicked should move to the top of the window (but not beneath the floating toolbars), assuming there are enough lines in the file that you can't see all of them at once. The line should also be highlighted in yellow.
4. Since this is a file that has rotations, you should see a label that says "File rotations" and some buttons next to it. Possible button labels are "Current" as well as any single-digit number between `0` and `9`. In this case "Current" means the file that has not hit the size threshold yet, so new log entries could be added to it. The numbered ones are files that have been rotated, with lower numbers being more recent than higher numbers. Clicking one of the buttons should take you to that file.
5. Try deleting one of the log files by clicking the "Delete permanently" button in the upper right corner. You should see a confirmation dialog popup before the file gets deleted. If you click "Cancel", the file should not be deleted!

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add a new single file view to the Logs screen for viewing the contents of a log file.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

coreymckrill committed:
Don't use full log file names in the UI

coreymckrill committed:
Improve file path validation

coreymckrill committed:
Scaffold the single file view

coreymckrill committed:
Refactor File to use wp_filesystem, add stream method

coreymckrill committed:
Render log file lines in single file view

coreymckrill committed:
Make log lines selectable, retain anchor

coreymckrill committed:
Remove test script

coreymckrill committed:
Extra escaping

coreymckrill committed:
phpcs cleanup

coreymckrill committed:
Fix broken markup

coreymckrill committed:
Fix broken markup

coreymckrill committed:
Style tweaks

coreymckrill committed:
Remove unused styles

coreymckrill committed:
Revert to css line numbers so they arent selectable

coreymckrill committed:
Add a header wrapper element on log screen titles

coreymckrill committed:
Add file rotation navigation

coreymckrill committed:
Add assertions for get_file_id to File unit tests

coreymckrill committed:
Add unit tests for getting files by file ID

coreymckrill committed:
Add some padding to Logs screen headers

coreymckrill committed:
Add delete button to single file view

coreymckrill committed:
Add more unit tests for File class

coreymckrill committed:
Add unit test for get_file_rotations, fix delete_files test

coreymckrill committed:
Ensure list table screen options aren't registered for single file view

coreymckrill committed:
Add a confirmation dialog for deleting the viewed file

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

barryhughes commented:
The colorized severity levels look great to me, but I'm not sure they meet (WCAG) accessibility standards.

For instance, emergencies and alerts have background color `#ff4136` and forefround color `#ffffff`. [This tool](https://webaim.org/resources/contrastchecker/) suggests that (with our font size) this is a fail. However, the slightly darker `#EB0C00` seems to pass WCAG AA and the darker yet `#B30900` passes WCAG AAA (so, something we might consider is tweaking the actual colors to achieve better contrast).

Alternatively: what if we were to keep the current range of colors, but apply them differently. For instance, we could use them to form a border around some black-on-white text:

<div align="center"><img width="700" alt="althighlight" src="https://github.com/woocommerce/woocommerce/assets/3594411/6a81649c-93ff-415c-b602-916e673e5dcb"></div>

Or, we might use a `::before content` rule to add a colored marker of some kind immediately next to the severity label.

(I'm definitely not an accessibility expert, but still wanted to pause and check on this.)

barryhughes commented:
Non-essential thing ... I wonder if 'zebra striping' would further increase readability (ie, a very slightly different background color on alternating lines)?

barryhughes commented on plugins/woocommerce/src/Internal/Admin/Logging/PageController.php:
```diff
@@ -305,28 +463,67 @@ private function handle_list_table_bulk_actions(): void {
-					</div>
-					<?php
-				}
+		$level_severities = range( 100, 800, 100 );
+		$severity_levels  = array();
+		foreach ( $level_severities as $severity ) {
+			$severity_levels[] = WC_Log_Levels::get_severity_level( $severity );
+		}
```
Not essential we change this, but rather than figure this out here, what are your thoughts on adding a new method to `WC_Log_Levels` such as `get_severity_levels()` (plural), that basically returns (currently protected property) `WC_Log_Levels::$severity_to_level`?

coreymckrill replied:
That's a great idea 5fee9226ae66afb981f6d2346dae56fb14178d12

barryhughes commented on plugins/woocommerce/client/legacy/css/admin.scss:
```diff
@@ -1305,61 +1305,161 @@ table.wc_status_table--tools {
+ *   critical, error  -> orange
+ *   warning, notice  -> yellow
+ *   info             -> blue
+ *   debug            -> gree
```
```suggestion
 *   debug            -> green
```

coreymckrill committed:
Update plugins/woocommerce/client/legacy/css/admin.scss

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

barryhughes commented on plugins/woocommerce/src/Internal/Admin/Logging/PageController.php:
```diff
@@ -305,28 +463,67 @@ private function handle_list_table_bulk_actions(): void {
+		}
+
+		$classes = implode( ' ', $classes );
+		$line    = sprintf(
+			'<span id="L%1$d" class="%2$s">%3$s%4$s</span>',
+			absint( $line_number ),
+			esc_attr( $classes ),
+			sprintf(
+				'<a href="#L%1$d" class="line-anchor"></a>',
+				absint( $line_number )
+			),
+			sprintf(
+				'<span class="line-content">%s</span>',
+				wp_kses_post( $text )
+			)
+		);
+
+		return $line;
```
Also non-essential, just something to consider (remove an unnecessary assignment):

```suggestion

		return sprintf(
			'<span id="L%1$d" class="%2$s">%3$s%4$s</span>',
			absint( $line_number ),
			esc_attr( $classes ),
			sprintf(
				'<a href="#L%1$d" class="line-anchor"></a>',
				absint( $line_number )
			),
			sprintf(
				'<span class="line-content">%s</span>',
				wp_kses_post( $text )
			)
		);
```

barryhughes commented:
> Try deleting one of the log files by clicking the "Delete permanently" button in the upper right corner. You should see a confirmation dialog popup before the file gets deleted. If you click "Cancel", the file should not be deleted!

This works, but I wonder if we could smooth things out some more. After deleting a rotation, I see:

<div align="center"><img width="700" alt="Mostly empty page, with an error message stating 'This file does not exist.' Also, an information notice stating '1 log file deleted.'" src="https://github.com/woocommerce/woocommerce/assets/3594411/407f3304-5c41-45f8-9288-b66004117f3b"></div>

I wondered, would it be a slightly smoother experience to take them back to the file list (dropping the file-does-not-exist warning, but keeping the log-file-deleted confirmation)?

barryhughes commented on plugins/woocommerce/src/Internal/Admin/Logging/PageController.php:
```diff
@@ -161,13 +164,129 @@ private function render_file_list_page( array $params = array() ): void {
+			<h2>
+				<?php
+				printf(
+				// translators: %s is the name of a log file.
```
```suggestion
					// translators: %s is the name of a log file.
```

I'm not sure why this is so common, my own IDE seems to often do the same thing 🤔

barryhughes reviewed (commented):
This is great!

Left some questions and thoughts, but nothing major. Nice work 🎉

coreymckrill committed:
Adjust severity level colors to meet WCAG AAA requirements

In order to do this, but still have differentiated bright colors,
this changes it from a background color to a border color. Also tweaks
the border styles to make it look like a label icon.

coreymckrill commented:
> The colorized severity levels look great to me, but I'm not sure they meet (WCAG) accessibility standards.

This is a good call-out. I was just using the same CSS rules that were already in place for the severity level colors in the DB logs table. I tried using that WCAG tool to get different colors, but everything ended up basically being a dark brown color, so I took your other suggestion and gave the labels a colorized border only, while keeping the background of the text white.

![severity-labels](https://github.com/woocommerce/woocommerce/assets/916023/18662c38-79d5-4e60-81dc-3d9005981cdb)

Updated in 8e177ea441. This change will affect the DB logs view as well.

coreymckrill committed:
Highlight the currently selected line

coreymckrill commented:
> I wonder if 'zebra striping' would further increase readability

I thought about this, but I think it would only make sense if you alternated the color based on _log entry_ rather than _line_. I.e., some log entries are multiple lines. I spent some time trying to make this work with an CSS-only implementation, but I have not been successful so far.

One approach would be to use the `format_line` method to add some classes to the lines based on whether a previous line was the start of an entry (i.e. contains the timestamp at the beginning). However, I'd really like to avoid having that method have to remember any kind of "state" beyond the current line.

So for now, I'll probably skip the zebra striping. I did add some styling in 2a7a25c4ba1f9c37d7799ddc18a4b933c8df2261 so that the currently selected line has a highlight color.

coreymckrill committed:
Simplify retrieval of severity levels

coreymckrill committed:
Update plugins/woocommerce/src/Internal/Admin/Logging/PageController.php

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

coreymckrill committed:
Update plugins/woocommerce/src/Internal/Admin/Logging/PageController.php

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

coreymckrill committed:
Go to file list after deleting file from single view

coreymckrill commented:
> I wondered, would it be a slightly smoother experience to take them back to the file list (dropping the file-does-not-exist warning, but keeping the log-file-deleted confirmation)?

I tried to make this work before, but was having trouble. However, I just thought of a way to do it in a32bada820057795331eeb29ba2df702a4ad47a6

coreymckrill committed:
phpcs cleanup

barryhughes commented on plugins/woocommerce/client/legacy/css/admin.scss:
```diff
@@ -1305,61 +1305,185 @@ table.wc_status_table--tools {
+	border-bottom-left-radius: 0.5em;
 
+	&:before {
+		content: "•";
```
Ahh! I was curious as to how you achieved this effect.

barryhughes commented on plugins/woocommerce/src/Internal/Admin/Logging/FileV2/File.php:
```diff
@@ -44,16 +46,38 @@ class File {
 	public function __construct( $path ) {
+		global $wp_filesystem;
+		if ( ! $wp_filesystem instanceof WP_Filesystem_Direct ) {
+			WP_Filesystem();
```
I was wondering if we should handle a case where this returns false or null (imagine a 'rogue' filter on `filesystem_method` or `filesystem_method_file`) ... but perhaps the answer is that there isn't much we could do in such a situation?

coreymckrill replied:
Good question. `WP_Filesystem()` does return `false` if something goes wrong, so we could capture that, but I'm not sure what we'd do with it... at that point it might be best to just let an error get triggered when we try to access a method on `$wp_filesystem` that doesn't exist.

I added [a card](https://github.com/orgs/woocommerce/projects/184/views/1?pane=issue&itemId=44071475) to the project board to think more about this.

barryhughes reviewed (approved):
This is lovely, thank you for all the updates and attention-to-detail!

Hitting approve, but also leaving a couple more thoughts that would probably constitute scope creep if we implemented them now (but I wanted to record them anyway and give you a chance to look them over)—more than happy for us to move ahead, though.

---

So one thing is that the colored severity labels work especially well when there are multiline errors, or backtraces, etc, as they make it really easy to locate the first line of the error:

<div align="center"><img width="700" alt="A multi-line error log entry, with colored severity label." src="https://github.com/woocommerce/woocommerce/assets/3594411/ab3d4f34-5611-4d6a-8beb-070d7ecc3d79"></div>

At the same time, with many single-line errors, it can become a bit "busy":

<div align="center"><img width="700" alt="Many single-line log entries, each with their own colored severity label." src="https://github.com/woocommerce/woocommerce/assets/3594411/107d7ea3-ff08-4472-96f9-fabcb0e89ada"></div>

The above is a pretty subjective, though. 

Lastly, I wondered about possible extensions including a *back to list of log files* link (I realize users can simply click on the *Logs* tab, but wonder if that's immediately intuitive to everyone) and perhaps an option to delete all logs in a given set?

<div align="center"><img width="700" alt="Amended view of the error log viewer, with additional links." src="https://github.com/woocommerce/woocommerce/assets/3594411/4c4eca68-ea38-42de-a64c-001ced12ef70"></div>

Again, all of the above are optional/not required.

coreymckrill commented:
@barryhughes some good additional thoughts about the UI. I created [a card](https://github.com/orgs/woocommerce/projects/184/views/1?pane=issue&itemId=44071930) on the project board to collect ideas for additional UI tweaks. They can be done in a separate PR, but I'd rather just get this one merged for now.

coreymckrill commented:
Rebasing to try and solve the seemingly unrelated unit test failures...

barryhughes commented:
> They can be done in a separate PR, but I'd rather just get this one merged for now.

Agree, sounds great :+1:
